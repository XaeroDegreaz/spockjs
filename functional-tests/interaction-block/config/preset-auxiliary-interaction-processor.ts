import { InteractionProcessor } from '@spockjs/config';

export const assertionPostProcessors = [];

export const interactionProcessors: InteractionProcessor[] = [
  t => ({
    primary: false,
    declare: () => t.emptyStatement(),
    verify: () => t.emptyStatement(),
  }),
];
