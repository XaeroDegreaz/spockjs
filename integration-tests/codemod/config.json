{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert"
      }
    ]
  ]
}
