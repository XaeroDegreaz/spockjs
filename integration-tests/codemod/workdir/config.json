{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert",
        "presets": ["@spockjs/preset-sinon-mocks"]
      }
    ]
  ]
}
