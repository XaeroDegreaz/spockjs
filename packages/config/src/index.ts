import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

import { Hooks } from './hooks';

/**
 * Config as it can be provided by the user.
 * Comments indicate transformations made to conform to InternalConfig format.
 */
export const configType = t.exact(
  t.partial({
    powerAssert: t.boolean,
    autoImport: t.union([t.boolean, t.string]),
    staticTruthCheck: t.boolean,
    assertFunctionName: t.string,
    presets: t.array(t.string),
  }),
  'Config',
);
export type Config = t.TypeOf<typeof configType>;

/**
 * Config that disables all but the core set of features.
 */
export const minimalConfig: Config = {
  powerAssert: false,
  autoImport: false,
};

/**
 * Config containing the settings used if they are missing in the user-specified config.
 */
export const defaultConfig = {
  powerAssert: true,
  autoImport: true,
  staticTruthCheck: false,
  assertFunctionName: '',
  presets: [],
};

/**
 * Config after sanitizing in extractConfigFromState.
 */
export interface InternalConfig {
  readonly powerAssert: boolean;
  readonly autoImport: string; // empty => no auto import
  readonly staticTruthCheck: boolean;
  readonly assertFunctionName: string; // empty => no custom assert function name
  readonly hooks: Hooks;
}

export const extractConfigFromState = ({
  opts,
}: {
  opts: any;
}): InternalConfig => {
  const decoded = configType.decode(opts);

  let {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    presets,
  } = decoded.getOrElseL(errors => {
    throw new Error(
      ['Invalid plugin configuration:', ...failure(errors)].join('\n'),
    );
  });

  // powerAssert
  if (powerAssert === undefined) {
    ({ powerAssert } = defaultConfig);
  }

  // autoImport
  if (autoImport === undefined) {
    ({ autoImport } = defaultConfig);
  }
  if (autoImport === false) {
    autoImport = '';
  }
  if (autoImport === true) {
    autoImport = 'power-assert';
  }

  // staticTruthCheck
  if (staticTruthCheck === undefined) {
    ({ staticTruthCheck } = defaultConfig);
  }

  // assertFunctionName
  if (assertFunctionName === undefined) {
    ({ assertFunctionName } = defaultConfig);
  }

  // presets
  if (presets === undefined) {
    ({ presets } = defaultConfig);
  }
  const hooks = presets.map(preset => require(preset) as Hooks).reduce(
    (accHooks, moreHooks) => ({
      assertionPostProcessors: [
        ...accHooks.assertionPostProcessors,
        ...moreHooks.assertionPostProcessors,
      ],
    }),
    { assertionPostProcessors: [] },
  );

  return {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    hooks,
  };
};
