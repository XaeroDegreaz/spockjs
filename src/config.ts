export interface Config {
  powerAssert: boolean;
}

const defaultConfig: Config = {
  powerAssert: true,
};

export default defaultConfig;
