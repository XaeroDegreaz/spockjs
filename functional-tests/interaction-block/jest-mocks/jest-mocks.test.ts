import { transform, TransformOptions } from '@babel/core';

import plugin from '@spockjs/babel-plugin-spock';
import { Config, minimalConfig } from '@spockjs/config';

// mark implicit dependency for jest
() => require('@spockjs/preset-jest-mocks');

const babelOpts: TransformOptions = {
  plugins: [
    [
      plugin,
      {
        ...minimalConfig,
        presets: ['@spockjs/preset-jest-mocks'],
      } as Config,
    ],
  ],
  filename: 'test.js',
};

test('reports too few calls', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 2 * mock();

    mock();

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('reports no calls at all', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 1 * mock();

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('reports too many calls', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 0 * mock();

    mock();

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('does not accept a call with missing arguments', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 1 * mock(42);

    mock();

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('does not accept a call with incorrect arguments', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 1 * mock(42);

    mock(1337);

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('does not throw if the calls were correct', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 2 * mock(42);

    mock(42);
    mock(42);

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).not.toThrowError();
});

test('does not care about unspecified calls', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 1 * mock(42);

    mock(42);
    mock(1337);

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).not.toThrowError();
});

test('does not care about mocks without interaction declarations', () => {
  const { code } = transform(
    `const mock = jest.fn();

    mock(42);
    mock(1337);

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).not.toThrowError();
});

test('reports too few calls with one of multiple argument lists', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 2 * mock();
    mock: 2 * mock(42);

    mock();
    mock();
    mock(42);

    verify: mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('works with a member expression mock', () => {
  const { code } = transform(
    `const mock = { method: jest.fn() };
    mock: 2 * mock.method();

    mock.method();

    verify: mock.method;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('works with a computed mock object', () => {
  const { code } = transform(
    `const mock = jest.fn();
    mock: 2 * mock();

    mock();

    verify: 0 || mock;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('throws if the callee in an interaction declaration is not a Jest mock function', () => {
  const { code } = transform(
    `const fn = () => {};
    mock: 2 * fn();`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});

test('throws if the value in an interaction verification is not a Jest mock function', () => {
  const { code } = transform(
    `const fn = () => {};
    verify: fn;`,
    babelOpts,
  );
  expect(() =>
    new Function('jest', 'require', code as string)(jest, require),
  ).toThrowErrorMatchingSnapshot();
});
