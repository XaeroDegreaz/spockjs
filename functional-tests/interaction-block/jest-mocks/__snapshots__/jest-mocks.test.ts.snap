// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not accept a call with incorrect arguments 1`] = `
"Expected 1 call(s) to mock jest.fn() with arguments Array [
  42,
], but received 0 such call(s)."
`;

exports[`does not accept a call with missing arguments 1`] = `
"Expected 1 call(s) to mock jest.fn() with arguments Array [
  42,
], but received 0 such call(s)."
`;

exports[`reports no calls at all 1`] = `"Expected 1 call(s) to mock jest.fn() with arguments Array [], but received 0 such call(s)."`;

exports[`reports too few calls 1`] = `"Expected 2 call(s) to mock jest.fn() with arguments Array [], but received 1 such call(s)."`;

exports[`reports too few calls with one of multiple argument lists 1`] = `
"Expected 2 call(s) to mock jest.fn() with arguments Array [
  42,
], but received 1 such call(s)."
`;

exports[`reports too many calls 1`] = `"Expected 0 call(s) to mock jest.fn() with arguments Array [], but received 1 such call(s)."`;

exports[`throws if the callee in an interaction declaration is not a Jest mock function 1`] = `"Expected the callee in an interaction declaration to be a Jest mock function, but received [Function fn]."`;

exports[`throws if the value in an interaction verification is not a Jest mock function 1`] = `"Expected the value in an interaction verification to be a Jest mock function, but received [Function fn]."`;

exports[`works with a computed mock object 1`] = `"Expected 2 call(s) to mock jest.fn() with arguments Array [], but received 1 such call(s)."`;

exports[`works with a member expression mock 1`] = `"Expected 2 call(s) to mock jest.fn() with arguments Array [], but received 1 such call(s)."`;
